### PR TITLE
Fix bug with aliased column in Distributed table

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix bug with aliased column in `Distributed` table

Introduced in https://github.com/ClickHouse/ClickHouse/pull/24405

cc @amosbird, could you take a look?

### CODE REVIEW

File: tests/queries/0_stateless/02008_aliased_column_distributed_bug.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+DROP TABLE IF EXISTS click_storage;
+DROP TABLE IF EXISTS click_storage_dst;
+
+CREATE TABLE click_storage ( `PhraseID` UInt64, `PhraseProcessedID` UInt64 ALIAS if(PhraseID > 0, PhraseID, 0) ) ENGINE = MergeTree() ORDER BY tuple();
+
+CREATE TABLE click_storage_dst ( `PhraseID` UInt64, `PhraseProcessedID` UInt64 ) ENGINE = Distributed(test_shard_localhost, currentDatabase(), 'click_storage');
+
+SELECT materialize(PhraseProcessedID)　FROM click_storage_dst;
```
Comment (amosbird) [inline tests/queries/0_stateless/02008_aliased_column_distributed_bug.sql:8]: We need to test both local plans and remote plans. In order to achieve this, some data has to be inserted. Then we use `settings prefer_localhost_replica = 0` and `settings prefer_localhost_replica = 1`

File: src/Interpreters/TreeRewriter.cpp [hunk 2]
```diff
@@ -976,7 +976,7 @@ TreeRewriterResultPtr TreeRewriter::analyzeSelect(
     result.required_source_columns_before_expanding_alias_columns = result.required_source_columns.getNames();
 
     /// rewrite filters for select query, must go after getArrayJoinedColumns
-    if (settings.optimize_respect_aliases && result.metadata_snapshot)
+    if (settings.optimize_respect_aliases && result.metadata_snapshot && !select_options.hasShardInfo())
```
Comment (amosbird) [inline src/Interpreters/TreeRewriter.cpp:979]: So we cannot expand aliases when it's not an initial query. We have a better check for this:

https://github.com/ClickHouse/ClickHouse/blob/7ee2dc04a1f67963e7e32db9873031d325a0c2bb/src/Interpreters/ExpressionAnalyzer.cpp#L252

**Review by amosbird — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/02008_aliased_column_distributed_bug.sql (hunk 1 above)
Comment (vdimir) [reply tests/queries/0_stateless/02008_aliased_column_distributed_bug.sql:13]: When I added it it broke my solution. But with check of `distributed_depth` and `distributed_group_by_no_merge` that you suggested seems that it works, thanks!

Let's wait for CI.

File: src/Interpreters/SelectQueryOptions.h [hunk 3]
```diff
@@ -138,6 +138,8 @@ struct SelectQueryOptions
         shard_count = shard_count_;
         return *this;
     }
+
+    bool hasShardInfo() const { return shard_num.has_value() || shard_count.has_value(); }
```
Comment (amosbird) [inline src/Interpreters/SelectQueryOptions.h:142]: Can be removed.

File: src/Interpreters/TreeRewriter.cpp [hunk 4]
```diff
@@ -976,7 +976,8 @@ TreeRewriterResultPtr TreeRewriter::analyzeSelect(
     result.required_source_columns_before_expanding_alias_columns = result.required_source_columns.getNames();
 
     /// rewrite filters for select query, must go after getArrayJoinedColumns
-    if (settings.optimize_respect_aliases && result.metadata_snapshot)
+    bool is_initiator = getContext()->getClientInfo().distributed_depth == 0 || settings.distributed_group_by_no_merge > 0;
```
Comment (amosbird) [inline src/Interpreters/TreeRewriter.cpp:979]: I'm sorry that I didn't make it clear. `getContext()->getClientInfo().distributed_depth == 0 ` is enough. `settings.distributed_group_by_no_merge > 0` is used for other purposes and should not be applied here (it's wrong)

**Review by amosbird — CHANGES_REQUESTED** [review]

**Pull request merged**